### DISCUSSION TITLE
Azure DevOps i downgrade az-cli na obrazie windows

### DISCUSSION THREAD

**Discussion by kopendrex** [AzureCommunityPL/group#331] [Pomoc]
Czy jest możliwe uruchomienie na ADO konkretnej wersji az-cli? Ostatnia wypuszczona 2.27.0 ma problem z "az container exec" na image windows-latest a na linuxach nie działa nie wiadomo już ile (https://github.com/Azure/azure-cli/issues/13352). Wersja 2.26.1 działała najnowsza już nie. Jak to downgradować? Próbowałem choco ale install nie przechodzi.

**Comment by huczas** [comment]
Jak masz postawiony swój worker to chulaj dusza, instalujesz co chcesz

**Reply by kopendrex** [reply]
Jakiś link do przykładu/dokumentacji podrzucisz?

**Reply by huczas** [reply]
w sensie jak dodać swój worker do ADO? czy jak na Linuksie zainstalować konkretną wersję AZ Cli?

**Reply by kopendrex** [reply]
Swój worker do ADO.

**Reply by huczas** [reply]
https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/agents?view=azure-devops&tabs=browser#install

**Reply by huczas** [reply]
a tu masz stricte dla linux https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-linux?view=azure-devops

**Reply by kopendrex** [reply]
Niestety brak uprawnień by podążać tą ścieżką :/

**Reply by huczas** [reply]
no to workery MSu są nie do ruszenia. Obawiam się że jedyna możliwość to zdobycie uprawnień w ADO

**Reply by kamil-mrzyglod** [reply]
@huczas Nie do końca jest to prawdą. Hosted agents w ADO są zadziwiająco elastyczne i można je w wielu miejscach dowolnie rzeźbić. Minus taki, że trzeba to robić przy każdym uruchomieniu co często zabija koncept.

**Reply by huczas** [reply]
> @huczas Nie do końca jest to prawdą. Hosted agents w ADO są zadziwiająco elastyczne i można je w wielu miejscach dowolnie rzeźbić. Minus taki, że trzeba to robić przy każdym uruchomieniu co często zabija koncept. 

Tak, można ale jak wspomniałeś zabiera to cenny czas, w jednym z projektów dobijałem do limitu wykonywania pajpa przez właśnie takie kombinacje i najprostrze było postawienie swojej vmki w azure z konkretnymi wersjami narzędzi co po prostu przyspieszyło realizacje o jakieś 20%. No i na swoim workerze nie ma limitu czasowego :)

**Comment by miqm** [comment]
a probowałeś odpalić dockera i podmontować volume z bieżącym katalogiem z repo?

**Comment by mgrabarz** [comment]
Nie jest to może bezpośrednia odpowiedź i pewnie az cli jest wygodniejsze, ale to tylko narzędzie opakowujące rest API.

Jeśli więc nie uda ci się downgrade, zawsze zostaje PowerShell lub bezpośredni call do API: https://docs.microsoft.com/en-us/rest/api/container-instances/containers/execute-command ;)

**Reply by kopendrex** [reply]
Wygląda obiecująco, jeszcze prośba o jakąś podpowiedź jak z poziomu bash uzyskać odpowiedź z tego terminala wss który jest zwracany jako odpowiedź API.

**Comment by kamil-mrzyglod** [comment]
Ja tam nie wiem co Ci nie działa ;) Pisane na kolanie.

```
# Starter pipeline
# Start with a minimal pipeline that you can customize to build and deploy your code.
# Add steps that build, run tests, deploy, and more:
# https://aka.ms/yaml

trigger:
- master

pool:
  vmImage: ubuntu-latest

steps:
- script: |
    sudo apt-get update
    curl -sL https://packages.microsoft.com/keys/microsoft.asc |
    gpg --dearmor |
    sudo tee /etc/apt/trusted.gpg.d/microsoft.gpg > /dev/null
    AZ_REPO=$(lsb_release -cs)
    echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ $AZ_REPO main" |
    sudo tee /etc/apt/sources.list.d/azure-cli.list
    sudo apt-get update
    sudo apt install -y --allow-downgrades azure-cli=2.26.0-1~focal
  displayName: 'Run a one-line script'

- script: |
    az version
  displayName: 'Run a multi-line script'
```

Wynik działania:
```
========================== Starting Command Output ===========================
/usr/bin/bash --noprofile --norc /home/vsts/work/_temp/26bdabdb-c55b-48a0-a6b4-00a12e5f3e1e.sh
{
  "azure-cli": "2.26.0",
  "azure-cli-core": "2.26.0",
  "azure-cli-telemetry": "1.0.6",
  "extensions": {
    "azure-devops": "0.18.0"
  }
}
```

Inspiracja - https://stackoverflow.com/questions/60774769/how-to-set-up-az-cli-concrete-version-in-azure-devops

EDIT: No chyba że Windows ma jakieś bolączki z tym związane, ale może wymaga po prostu jakiejś rzeźby w przeciwieństwie do Linuxa.

**Reply by kopendrex** [reply]
Dokładnie tak, chodzi o to że na linuxach jest błąd od dłuższego czasu, który nie pozwala na az container exec (https://github.com/Azure/azure-cli/issues/13352) obejściem było używanie obrazu windows. Niestety wczoraj wjechał az-cli 2.27.0 który również failuje, na 2.26.1 działało stąd cała moja sytuacja.

**Reply by miqm** [reply]
Możesz dopisać to do tego błędu na githubie? Wygląda na coś czym zespół CLI powinien się szybko zająć

**Reply by kopendrex** [reply]
W wersji 2.27.0 na windows jest to inny błąd `OSError: [WinError 6] The handle is invalid` i raczej nie związany z wcześniejszym.

**Reply by miqm** [reply]
Jest na to issue?

**Reply by kopendrex** [reply]
Założyłem: https://github.com/Azure/azure-cli/issues/19133

**Reply by kamil-mrzyglod** [reply]
@kopendrex A możesz napisać jaki problem z choco miałeś? Jak będę miał chwilę to spróbuję zrobić downgrade na agencie Win.

**Reply by kopendrex** [reply]
choco nie może usunąć ani zainstalować az-cli, wymagany byłby restart maszyny.

**Reply by kamil-mrzyglod** [reply]
Fakt, może być takie issue. W takim razie chyba najlepszym rozwiązaniem byłoby pewnie tak jak Marek napisał - albo call bezpośrednio do API, albo Powershell - https://docs.microsoft.com/en-us/powershell/module/az.containerinstance/invoke-azcontainerinstancecommand?view=azps-6.3.0
